dali/kernels/imgproc/roi.h

```diff
@@ -0,0 +1,158 @@
+namespace kernels {
+
+/**
+ * Defines region of interest.
+ * 0 dimension is interpreted along x axis (horizontal)
+ * 1 dimension is interpreted along y axis (vertical)
+ *
+ *            image.x ->
+ *          +--------------------------------+
+ *          |                                |
+ *          |   roi.lo    roi.x              |
+ *  image.y |         +-----+                |
+ *       |  |    roi.y|     |                |
+ *       v  |         +-----+ roi.hi         |
+ *          |                                |
+ *          +--------------------------------+
+ *
+ * Additionally, by definition, ROI is top-left inclusive and bottom-right exclusive.
+ * That means, that `Roi.lo` point is included in actual ROI and `Roi.hi` point isn't.
+ */
+template <size_t ndims>
+using Roi = Box<ndims, int>;
+
+namespace detail {
+
+/**
+ * Create Roi for whole image defined by TensorShape
+ */
+template <size_t spatial_dims>
+Roi<spatial_dims> whole_image(const TensorShape<spatial_dims + 1> &shape) {
+  ivec<spatial_dims> size;
+  for (size_t i = 0; i < spatial_dims; i++)
+    size[i] = shape[spatial_dims - 1 - i];
+  return {0, size};
+}
+
+}  // namespace detail
+
+
+/**
+ * Defines TensorShape corresponding to provided Roi.
+ *
+ * Function assumes, that memory layout is *HWC, and the Roi
+ * is represented as [[x_lo, y_lo], [x_hi, y_hi]].
+ * Therefore, while copying, order of values needs to be reversed.
+ *
+ * @tparam ndims Number of dims in Roi
+ * @param roi Region of interest
+ * @param nchannels Number of channels in data
+ * @return Corresponding TensorShape
+ */
+template <size_t ndims>
+TensorShape<ndims + 1> ShapeFromRoi(const Roi<ndims> &roi, int nchannels) {
+  DALI_ENFORCE(all_coords(roi.hi >= roi.lo), "Cannot create a TensorShape from an invalid Roi");
+  TensorShape<ndims + 1> ret;
+  auto e = roi.extent();
+  auto ridx = ndims;
+  ret[ridx--] = nchannels;
+  for (size_t idx = 0; idx < ndims; idx++) {
+    ret[ridx--] = e[idx];
+  }
+  return ret;
+}
+
+
+/**
+ * Convenient overload for batch processing (creating TensorListShape)
+ */
+template <size_t ndims>
+TensorListShape<ndims + 1> ShapeFromRoi(const std::vector<Roi<ndims>> &rois, int nchannels) {
+  DALI_ENFORCE(!rois.empty(), "Provided argument doesn't contain any Roi");
+  TensorListShape<ndims + 1> ret(rois.size());
+  size_t i = 0;
+  for (const auto &roi : rois) {
+    ret.set_tensor_shape(i++, ShapeFromRoi(roi, nchannels));
+  }
+  return ret;
+}
+
+
+/**
+ * Adjusted Roi is a Roi, which doesn't overflow the image, that is given by TensorShape.
+ *
+ * If `rois` is not provided (roi == nullptr), that means whole image is analysed:
+ * return a Roi, that has the same size as the input image
+ *
+ * If `rois` is provided, adjusted Roi is an intersection of provided Roi and the image.
+ *
+ * Assumes HWC memory layout
+ */
+template <size_t spatial_dims>
+Roi<spatial_dims>
+AdjustRoi(const Roi<spatial_dims> *roi, const TensorShape<spatial_dims + 1> &shape) {
```

Comment:
I'd stick to same template argument `ndims` as total number of dimensions and you can define `spatial_dims` as `ndims - 1` when necessary

Reply:
Done